splink/internals/linker_components/clustering.py

```diff
@@ -334,6 +346,307 @@ def cluster_using_single_best_links(
+            c.unquote().name for c in df_predict.columns
+        ]
+
+        is_match_weight = (
+            threshold_match_weights is not None
+            and threshold_match_probabilities is None
+        )
+
+        threshold_match_probabilities = threshold_args_to_match_prob_list(
+            threshold_match_probabilities, threshold_match_weights
+        )
+
+        if (
+            threshold_match_probabilities is None
+            or len(threshold_match_probabilities) == 0
+        ):
+            raise ValueError(
+                "Must provide either threshold_match_probabilities "
+                "or threshold_match_weights"
+            )
+
+        if not has_match_prob_col and threshold_match_probabilities is not None:
+            raise ValueError(
+                "df_predict must have a column called 'match_probability' if "
+                "threshold_match_probability is provided"
+            )
+
+        initial_threshold = threshold_match_probabilities.pop(0)
+        all_results = {}
+        all_results_summary = {}
+
+        match_p_expr = ""
+        match_p_select_expr = ""
+        if initial_threshold is not None:
+            match_p_expr = f"where match_probability >= {initial_threshold}"
+            match_p_select_expr = ", match_probability"
+
+        pipeline = CTEPipeline([df_predict])
+
+        # Templated name must be used here because it could be the output
+        # of a deterministic link i.e. the physical name is not know for sure
+        sql = f"""
+        select
+            {uid_concat_edges_l} as node_id_l,
+            {uid_concat_edges_r} as node_id_r
+            {match_p_select_expr}
+            from {df_predict.templated_name}
+            {match_p_expr}
+        """
+        pipeline.enqueue_sql(sql, "__splink__df_edges_from_predict")
+
+        edges_table_with_composite_ids = db_api.sql_pipeline_to_splink_dataframe(
+            pipeline
+        )
+        logger.info(f"--------Clustering at threshold {initial_threshold}--------")
+        # First cluster at the lowest threshold
+        cc = cluster_pairwise_predictions_at_threshold(
+            nodes=nodes_with_composite_ids,
+            edges=edges_table_with_composite_ids,
+            db_api=db_api,
+            node_id_column_name="node_id",
+            edge_id_column_name_left="node_id_l",
+            edge_id_column_name_right="node_id_r",
+            threshold_match_probability=initial_threshold,
+        )
+
+        all_results[initial_threshold] = cc
+
+        # Calculate Summary stats for first clustering threshold
+        pipeline = CTEPipeline([cc])
+        sqls = _get_cluster_stats_sql(cc)
+        pipeline.enqueue_list_of_sqls(sqls)
+        cc_summary = db_api.sql_pipeline_to_splink_dataframe(pipeline)
+        all_results_summary[initial_threshold] = cc_summary
+
+        # Now iterate over the remaining thresholds
+        previous_threshold = initial_threshold
```

Comment:
Can we reuse cluster_pairwise_predictions_at_multiple_thresholds [here](
https://github.com/moj-analytical-services/splink/blob/d1326c201de011e751ab515af4082f39e7da026e/splink/internals/clustering.py#L346) rather than reimplementing the logic?